### ISSUE TITLE
@def not defined

### ISSUE THREAD

**Issue by joa-quim** [tlienart/Franklin.jl#964]
Hi, I'm struggling to deploy may ``__site``. For example, the docs say

>In any case, before deploying, if you're working on a project website i.e. a website whose root URL will look like username.gitlab.io/project/ then you should add the following line in your config.md file:
>
>@def prepath = "project"

But if I do it
```
ERROR: Encountered an error on  while trying to evaluate a block of definitions (`+++...+++`).
Stacktrace:
...
   @ REPL[4]:1

caused by: LoadError: UndefVarError: @def not defined
```

**Comment by tlienart** [comment]
Hello @joa-quim, 

Assuming that you're deploying your website off GitHub and that the repository is `joa-quim/TheRepo` what you need to indicate is that the prefix is `TheRepo` i.e. the landing page will be at `joa-quim.github.io/TheRepo/` 

To do so, the easiest is to set `prepath = "TheRepo"` in your `config.md` file. 

The `config.md` file can have definitions following two ways: (1) the `@def` way (which is being deprecated) and (2) within a `+++...+++` block. Both of the following would work for `config.md`:

```
+++
# potentially other definitions here
prepath = "TheRepo"
# potentially more definitions after
+++
```

or

```
@def prepath = "TheRepo"
@def someOtherVar = 5
```

you can't mix the two (which is what I think you did here) so this is not ok:

```
+++
@def prepath = "TheRepo"
+++
```

I hope that helps you solve the problem, if not please indicate your repo (or better yet add me as collaborator).

**Comment by joa-quim** [comment]
Thanks for quick answer and kind offer to assist, which I will quite likely accept but I'm not at that stage yet.
I intend to hos the site in Github. In fact I intend to replace the [GMT.jl](https://www.generic-mapping-tools.org/GMT.jl/dev/) docs that are giving a non-stop flow of headaches. But so far I only have the trial site on my computer.

I did try setting ``prepath = "GMTjl_docs"`` as well as ``optimize(prepath="GMTjl_docs", minify=false)`` but if I grep the entire ``__site`` dir for ``GMTjl_docs`` nothing shows up. I have a site with an html server that I intended to use as a test but the result is miserable. The several ``indext.html`` have no reference to the location ``GMTjl_docs`` directory.

 http://fct-gmt.ualg.pt/GMTjl_docs/

**Comment by joa-quim** [comment]
OK, so I created a new [repo ](https://github.com/joa-quim/GMTjl_doc)and added you as collaborator. Oddly the ``__site`` directory does not seem to have been uploaded. At least I don't see it (we should see it, no?).
I don't know how this is supposed to work. Is some action triggered that runs ``repo()``? Because if yes, or you try to do it manually, installing GMT in Linux seems to be currently broken because of Conda playing newest-of-newest compilers game (see this [forum post](https://discourse.julialang.org/t/failing-to-register-new-version-because/82346/8)).

So basically I'm back to the beginning. No deployment.

**Comment by tlienart** [comment]
Alright, it's night time here so let me have a look tomorrow and see what I can do to unblock you, I think it should be easy for the most part (not sure about your anaconda comment but I guess that will make more sense once I've had a look at the repo)

**Comment by joa-quim** [comment]
Sure, it's night time here too (GMT+1) but I'm a night guy.

**Comment by joa-quim** [comment]
I've spend many hours around this and, so far, have to conclude the problem is really with the ``prepath``. It never gets set. Copying the contents of ``__site`` to the root of an html server works. *e.g.* http://fct-gmt.ualg.pt/
but copying it into a directory, no matter what I try with ``optimize(prepath="GMTjl_docs", minify=false)`` doesn't. *e.g.* http://joa-quim.pt/GMTjl_docs/

If I do ``optimize(prepath="GMTjl_docs", version="1.43", minify=false)`` then it creates copies over copies of ``__site`` under ``dev``, ``stable`` and ``1.43``.

Trying Github pages didn't work either. Just don't get on how to put the site in the ``gh-pages``, either manually or a tentative actions script.

**Comment by tlienart** [comment]
Dont waste more time on it (Sorry for the frustration), it's super easy (I mean there's hundreds of Franklin websites out there) I just need to get a bit of time to have a look and help you fix it, maybe there's a small mistake somewhere that keeps messing things up for you. I just haven't had the time to look into it yet.

**Comment by tlienart** [comment]
did you write the `makedocs.jl` or did you copy it from somewhere else? 

that is most likely what messes things up, I suggest removing this, using a common deployment workflow like this one https://github.com/tlienart/FranklinTemplates.jl/blob/master/src/templates/common/.github/workflows/Deploy.yml unless you have very good reasons not to.

**Comment by tlienart** [comment]
if you want `lunr `on top of that you would add to that script:

1. the installation of the lunr stuff (https://github.com/tlienart/Franklin.jl/blob/ed861a3fe0ed04ec44db76046d147feec9b392ff/.github/workflows/deploy.yml#L31-L33)
2. replace the single optimize call with this: https://github.com/tlienart/Franklin.jl/blob/ed861a3fe0ed04ec44db76046d147feec9b392ff/.github/workflows/deploy.yml#L38

**Comment by joa-quim** [comment]
I am adapting from the Makie site and ``makedocs.jl`` was written by them. And in fact that's what made things work. Before I was using ``serve()`` with the problems I related. Now, by running ``makedocs.jl`` I can finally get a ``__site`` directory that I can copy to an html server. ... except the damn gh-pages that I still didn't figure out how to make work. The github actions finish, say they are uploading the artifact  but nothing shows up in gh-pages.
But I'll do the changes that you say.

**Comment by tlienart** [comment]
ok so unfortunately I don't manage to build your website locally, there's an issue with something called `set_config_option`.

Could I suggest starting from scratch in a simple repo with just one or two pages that can be built locally, removing any non-essential utils stuff and maybe having only a single basic figure built with GMT, so that we just get something deployed properly, and then add the stuff back in.

Also, since you're starting this from scratch, it might be worth it to directly go with Xranklin (the upcoming release of Franklin, I can do that myself though if you provide me with a minimal repo with just the basic stuff, I can adapt it to be ready for Xranklin).

**Comment by joa-quim** [comment]
Ok, I'll prepare that simple repo. But you mean one repo that just uses one GMT fig and that is not a striped down version of this, Makie based, one?
I liked very much the look of it with the tilled images that lead to sub-pages with more tilled images and I would like to have one with these features but ofc it doesn't need to be based in a complicated solution.

Regarding the errors you got. Quite likely you are not using GMT.jl 0.42.3 because those are errors that I remember to fix recently (exactly because creating those figures made them surface).

**Comment by tlienart** [comment]
> Ok, I'll prepare that simple repo. But you mean one repo that just uses one GMT fig and that is not a striped down version of this, Makie based, one?
> I liked very much the look of it with the tilled images that lead to sub-pages with more tilled images and I would like to have one with these features but ofc it doesn't need to be based in a complicated solution.

yes, as simple as possible, not caring about the final look, this is about getting a deployed site that you can increment on.

Tiled stuff is trivial to get with some CSS.

**Comment by tlienart** [comment]
ok after updating, I was actually able to build your website locally, so let me try to deploy it somewhere else

**Comment by tlienart** [comment]
ok this is good: https://github.com/joa-quim/GMTjl_doc/runs/6834625961?check_suite_focus=true the hardest was to install GMT, their installation instructions are a bit crap

you'll need to fix `coach.md` which has errors with relinks, probably you have stray braces, I didn't have time to look into this, just serve locally and uncomment it gradually until you fix the issues.

that should be a good starting point. I think this issue can be closed.

**Comment by joa-quim** [comment]
Wonderful, I'll dive on this and give news (hopefully the best) when ~finish

>the hardest was to install GMT, their installation instructions are a bit crap

Ugh, I'm GMT developer too (but more on the Windows side). We try make it as simple as possible but with all those Linux alternatives, different package managers, extremely old versions by default on the Ubuntu's words and not to mention the troubles Xcode creates on us so that the MacOS bundle has shared libraries that can be used by externals (Julia, Matlab, Python) it's hard to make it better. And, if my comments on this PR have escaped you, it has more *excuses*.

https://github.com/joa-quim/GMTjl_doc/commit/6b7ed8be9acbfcdf5cee0c37f82fac817fa40b63

**Comment by joa-quim** [comment]
Ah, and the curly braces error. That is quite strange because it seems to occur depending on some past state. The same code is used in many other places but the is some (``coast.md``) where it causes that error.

**Comment by tlienart** [comment]
Ah sorry about that, I meant it more as a groan than a blame, I definitely can imagine that it's a pain to provide multi platform instructions.

I'm just used to GitHub actions with linux latest, couldn't get the latest GMT so turned to macos with home brew, it works but it's super slow. Maybe the windows way is fastest but I just wanted something that worked

**Comment by tlienart** [comment]
Re reflink and coast, I think the easiest way here is to bracket, un-comment gradually until something fails when you serve locally, and see what it is, if you can isolate it but are not sure why that fails, open an issue in the repo and I'll have a look

The site looks good btw, I'm glad it's deployed :)

If you want to set it on a dedicated server, you'll have to figure out what the prepath has to be (likely the empty string depends a bit on your setup if you're at root or not) and you'll have to place the content of __site after the optimise step yourself (e.g by scp or some other relevant approach) easier still is to just use GitHub for all this but depends what you prefer.

**Comment by tlienart** [comment]
Note on installation: I expected the conda way to work on Ubuntu but it doesn't, it installs fine but I don't know where it places the executable. If you figure that one out I think that would be the best way

**Comment by joa-quim** [comment]
No problem with the rant. I know that, but I can't avoid thinking that unix _easy-on-breaking_ attitude is to blame too. Conda used to work, and will again when the problems with libstdc++ are solved (see end of this [forum post](https://discourse.julialang.org/t/failing-to-register-new-version-because/82346/9)) ... but then they will resurrect one day.

I have the site on a [dedicated server](http://fct-gmt.ualg.pt/GMTjl_doc/) too. If I run the ``makedocs.jl`` and copy the contents of ``__site`` it works.

The ``reflink`` issue. I tried that already but it's proving very illusive. It happened in other ``documentation/.md`` files and then stop erroring. It's bizarre, it happens with both ``_assets/common_opts/opt_R.md`` and ``_assets/common_opts/opt_I.md`` that are included in many places (BTW, a big thank you for allowing including .md files. Documenter doesn't and I _@#$%&@#_ it so many times for that). The ``coast.md`` case is just one case where I left the reflink explicit and at some point it started erroring.

**Comment by joa-quim** [comment]
> Note on installation: I expected the conda way to work on Ubuntu but it doesn't, it installs fine but I don't know where it places the executable.

It's somewhere inside ``.julia\conda\3\...`` (I don't have in my Windows machine. For Windows it installs a MSVC build GMT)

**Comment by tlienart** [comment]
Ah no i meant I followed the conda instructions to install the gmt library (not GMT.jl) so the conda-forge stuff. The installation worked I just don't know where the `gmt` is, if it can be added to the path somehow then we can go back to ubuntu and avoid the huge overhead of homebrew

Re error: could you try changing the definition of the reflink command to something really dumb like "HERE" or whatever, that you know would always work and not call some other command, and check if Franklin chokes on something else, hopefully that might point you in the right direction

(I find issuing useful error messages hard to do and Franklin is no exception)

**Comment by joa-quim** [comment]
That may be a hopeless (at least quite difficult) task. Taking the Windows example, Conda spreads things around and one would need to know also where the dependencies are and what ``LD_LIBRARY_PATH`` to set, etc. And at the end it would have the same libstdc++ compat error that mentioned above when GMT.jl automatically installs GMT via Conda.

I'll try that thing with the reflink.

**Comment by joa-quim** [comment]
but I need a little help here. I don't understand what this does nor how can change it to just print something.

```
\newcommand{\myreflink}[1]{[!#1](\reflink{!#1})}
```

**Comment by tlienart** [comment]
The set of curly brace after [1] is the definition (same as latex) so if you remove the entire content and replace it with `reflink:#1` the behaviour simply becomes to print "reflink: xxx" where "xxx" is the link.

The aim of this is to bypass the nested reflink command which is a Franklin-defined command. 

The purpose is to simplify stuff and try to find out what exactly causes issues.

**Comment by joa-quim** [comment]
Opened [issue](https://github.com/joa-quim/GMTjl_doc/issues/8)
